risc0/cargo-risczero/src/commands/deploy.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright 2024 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::path::PathBuf;
+
+use anyhow::{Context, Result};
+use bonsai_sdk::{alpha::Client, API_KEY_ENVVAR, API_URL_ENVVAR};
+use clap::Parser;
+use risc0_build::BuildStatus;
+
+use crate::commands::build_guest;
+
+/// Reports the current version of this crate.
+pub const VERSION: &str = env!("CARGO_PKG_VERSION");
+
+/// `cargo risczero deploy`
+///
+/// NOTE: Requires Docker to be installed and running.
+#[derive(Parser)]
+pub struct DeployCommand {
+    /// Location of the Cargo.toml for the guest code.
+    ///
+    /// This path is relative to the current directory.
+    #[arg(long)]
+    pub manifest_path: PathBuf,
+
+    /// Feature flags passed to cargo.
+    #[arg(long, value_delimiter = ',')]
+    pub features: Vec<String>,
+
+    /// API URL for Bonsai.
+    #[arg(long)]
+    pub api_url: Option<String>,
+
+    /// API key for Bonsai.
+    #[arg(long)]
+    pub api_key: Option<String>,
+}
+
+impl DeployCommand {
+    pub fn run(&self) -> Result<()> {
+        // Instantiate client first to check for errors
+        let client = self.get_client()?;
+        self.deploy(client)?;
+
+        Ok(())
+    }
+
+    fn deploy(&self, client: Client) -> Result<()> {
+        // Ensure we have an up to date artifact before deploying
+        if let BuildStatus::Skipped = build_guest::build(&self.manifest_path, &self.features)? {
+            eprintln!("Build skipped, nothing to deploy.");
+            return Ok(());
+        }
```

Comment:
I opted to always compile the guest using reproducible builds as it seemed the right thing to do. Please let me know if it would be desirable to upload non-reproducible builds too.